### PR TITLE
*: remove vendor requirements

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:**
* remove dependency on `vendor/` by various parts of the SDK
* add `--vendor` to direct `operator-sdk new` to create one (if using `modules`). If `dep` is the dependency manager, `vendor/` is still used.
* add a dependency validation step, which builds the project to ensure all dependencies compile, and a `--skip-validation` option to skip this step.

**Motivation for the change:** Go modules do not require a `vendor/` directory to work, instead using a module cache elsewhere. This PR is part of a larger effort to fully support module functionality. See the [modules wiki](https://github.com/golang/go/wiki/Modules) for an overview of this functionality.

### CODE REVIEW

**Comment by estroz** [comment]
Note: this is blocked by #1470. Code generators depend on a `vendor/` dir being present in `master`, so this PR would break them.

File: doc/user-guide.md [hunk 1]
```diff
@@ -416,7 +416,7 @@ func main() {
 
 **NOTES:**
 
-* After adding new import paths to your operator project, run `go mod vendor` (or `dep ensure` if you set `--dep-manager=dep` when initializing your project) in the root of your project directory to fulfill these dependencies.
+* After adding new import paths to your operator project, run `go mod vendor` if using a `vendor/` directory (or `dep ensure` if you set `--dep-manager=dep` when initializing your project) in the root of your project directory to fulfill these dependencies.
```
Comment (LiliC) [inline doc/user-guide.md]: Hmm this reads like the user should run `go mod vendor` when using `dep`.

File: doc/sdk-cli-reference.md [hunk 2]
```diff
@@ -263,6 +262,9 @@ Scaffolds a new operator project.
 * `--helm-chart-repo` string - Chart repository URL for the requested helm chart
 * `--helm-chart-version` string - Specific version of the helm chart (default is latest version)
 * `--dep-manager` string - Dependency manager the new project will use (choices: "dep", "modules") (default "modules")
+* `--skip-git-init` - Do not init the directory as a git repository
+* `--vendor` - Use a vendor directory for dependencies. This flag only applies when --dep-manager=modules (the default). If --no-check is set, vendoring is skipped
```
Comment (LiliC) [inline doc/sdk-cli-reference.md]: Nit: 
```suggestion
* `--vendor` - Use a vendor directory for dependencies. This flag only applies when `--dep-manager=modules` (the default). If `--no-check` is set, vendoring is skipped
```

as we have that formatting a few lines above as well.

File: cmd/operator-sdk/new/cmd.go [hunk 3]
```diff
@@ -60,6 +60,8 @@ generates a skeletal app-operator application in $GOPATH/src/github.com/example.
 	newCmd.Flags().StringVar(&depManager, "dep-manager", "modules", `Dependency manager the new project will use (choices: "dep", "modules")`)
 	newCmd.Flags().BoolVar(&skipGit, "skip-git-init", false, "Do not init the directory as a git repository")
 	newCmd.Flags().StringVar(&headerFile, "header-file", "", "Path to file containing headers for generated Go files. Copied to hack/boilerplate.go.txt")
+	newCmd.Flags().BoolVar(&makeVendor, "vendor", false, "Use a vendor directory for dependencies. This flag only applies when --dep-manager=modules (the default). If --no-check is set, vendoring is skipped")
+	newCmd.Flags().BoolVar(&noCheck, "no-check", false, "Do not validate the resulting projects' structure and dependencies")
```
Comment (LiliC) [inline cmd/operator-sdk/new/cmd.go]: Nit: Maybe we can call this `skip-validate` in the same way we have `skip-git-init`.

File: cmd/operator-sdk/new/cmd.go [hunk 4]
```diff
@@ -436,3 +462,26 @@ func initGit() error {
...
+		opts := projutil.GoCmdOptions{
+			PackagePath: "./...",
+			Dir:         filepath.Join(projutil.MustGetwd(), projectName),
+		}
+		if err := projutil.GoBuild(opts); err != nil {
+			return err
+		}
+	}
+
+	log.Info("Check project successful.")
```
Comment (LiliC) [inline cmd/operator-sdk/new/cmd.go]: In case of another dep manager we just succeed? :)

File: test/e2e/memcached_test.go [hunk 5]
```diff
@@ -218,9 +218,9 @@ func TestMemcached(t *testing.T) {
 	}
 
 	t.Log("Pulling new dependencies with go mod")
-	cmdOut, err = exec.Command("go", "mod", "vendor").CombinedOutput()
+	cmdOut, err = exec.Command("go", "build", "./...").CombinedOutput()
 	if err != nil {
-		t.Fatalf("Command 'go mod vendor' failed: %v\nCommand Output:\n%v", err, string(cmdOut))
+		t.Fatalf("Pulling modules failed: %v\nCommand Output:\n%v", err, string(cmdOut))
```
Comment (LiliC) [inline test/e2e/memcached_test.go]: Not sure pulling modules is the only error we could be running into, we should just say `go build failed` instead.

File: cmd/operator-sdk/new/cmd.go (hunk 4 above)
Comment (estroz) [reply cmd/operator-sdk/new/cmd.go]: Yes. We can add other validators in the future in general and for `dep`. If you think I should move the log statements so they do not execute unless modules are used, I can.

File: doc/user-guide.md [hunk 6]
```diff
@@ -53,7 +53,7 @@ $ export GO111MODULE=on
 
 ##### Vendoring
 
-The Operator SDK uses [vendoring][go_vendoring] to supply dependencies to operator projects, regardless of the dependency manager. As with the above module mode constraint, we intend to allow use of dependencies [outside of `vendor`][module_vendoring] for projects in the near future.
+The Operator SDK uses [vendoring][go_vendoring] to supply dependencies to Go operator projects if the dependency manager is `dep`, or `modules` and the project was initialized with the `--vendor` flag set. If the operator's dependencies are managed with `modules` and `--vendor` is not set, calls to `go {build,clean,get,install,list,run,test}` by `operator-sdk` subcommands will use an external modules directory. Execute `go help modules` for more information.
```
Comment (joelanford) [inline doc/user-guide.md]:
```suggestion
The Operator SDK uses [vendoring][go_vendoring] to supply dependencies to Go operator projects if the dependency manager is `modules` and the project was initialized with the `--vendor` flag set or if the dependency manager is `dep` (which always uses vendoring).

By default, the operator's dependencies are managed with `modules` and `--vendor` is not set, so calls to `go {build,clean,get,install,list,run,test}` by `operator-sdk` subcommands will use an external modules directory. Execute `go help modules` for more information.
```

File: doc/user-guide.md [hunk 7]
```diff
@@ -444,30 +444,30 @@ func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Res
...
  }
- 
+
 //addFinalizer will add this attribute to the Memcached CR
 func (r *ReconcileMemcached) addFinalizer(reqLogger logr.Logger, m *cachev1alpha1.Memcached) error {
     if len(m.GetFinalizers()) < 1 && m.GetDeletionTimestamp() == nil {
         reqLogger.Info("Adding Finalizer for the Memcached")
         m.SetFinalizers([]string{"finalizer.cache.example.com"})
-    
+
```
Comment (joelanford) [inline doc/user-guide.md]: Given the more extensive changes to the finalizer example in #1524 (and since these are unrelated to this PR), want me to make sure I've got the spacing issues covered there?

File: internal/util/projutil/exec.go [hunk 8]
```diff
@@ -106,10 +114,15 @@ func getGeneralArgs(cmd string, opts GoCmdOptions) ([]string, error) {
 		if goModOn, err := GoModOn(); err != nil {
 			return nil, err
 		} else if goModOn {
-			bargs = append(bargs, "-mod=vendor")
+			if info, err := os.Stat("vendor"); err == nil && info.IsDir() {
+				bargs = append(bargs, "-mod=vendor")
```
Comment (joelanford) [inline internal/util/projutil/exec.go]: Is it okay to add this flag to _any_ `go` command?

Comment (estroz) [reply internal/util/projutil/exec.go]: Having a command string check is a good idea.

File: cmd/operator-sdk/new/cmd.go [hunk 9]
```diff
@@ -436,3 +462,26 @@ func initGit() error {
 	log.Info("Run git init done")
 	return nil
 }
+
+func validateProject() error {
+	log.Info("Validating project")
+	switch projutil.DepManagerType(depManager) {
+	case projutil.DepManagerGoMod:
+		// Run "go build ./..." to make sure all packages can be built
+		// currectly. From "go help build":
```
Comment (LiliC) [inline cmd/operator-sdk/new/cmd.go]:
```suggestion
		// correctly. From "go help build":
```

File: cmd/operator-sdk/new/cmd.go [hunk 10]
```diff
@@ -436,3 +462,26 @@ func initGit() error {
...
+		opts := projutil.GoCmdOptions{
+			PackagePath: "./...",
+			Dir:         filepath.Join(projutil.MustGetwd(), projectName),
+		}
+		if err := projutil.GoBuild(opts); err != nil {
+			return err
+		}
+	}
+
+	log.Info("Project validation successful.")
```
Comment (LiliC) [inline cmd/operator-sdk/new/cmd.go]: Nit: I would just move this before }, so on line 483, if that makes sense. But just a nit, rather than anything else. :)

File: hack/tests/e2e-ansible.sh [hunk 11]
```diff
@@ -133,9 +133,9 @@ then
     exit 1
 fi
 
-# Right now, SDK projects still need a vendor directory, so run `go mod vendor`
-# to pull down the deps specified by the scaffolded `go.mod` file.
-go mod vendor
+# Run `go build ./..` to pull down the deps specified by the scaffolded
```
Comment (joelanford) [inline hack/tests/e2e-ansible.sh]: Nit
```suggestion
# Run `go build ./...` to pull down the deps specified by the scaffolded
```

File: cmd/operator-sdk/new/cmd.go [hunk 12]
```diff
@@ -368,8 +379,13 @@ func verifyFlags() error {
...
 	}
 
-	if operatorType == projutil.OperatorTypeGo && (len(apiVersion) != 0 || len(kind) != 0) {
-		return fmt.Errorf("operators of type Go do not use --api-version or --kind")
+	if operatorType == projutil.OperatorTypeGo {
+		if len(apiVersion) != 0 || len(kind) != 0 {
+			return fmt.Errorf("operators of type Go do not use --api-version or --kind")
+		}
+		if !makeVendor && projutil.DepManagerType(depManager) == projutil.DepManagerDep {
+			return fmt.Errorf("--vendor cannot be used with --dep-manager=dep")
```
Comment (hasbro17) [inline cmd/operator-sdk/new/cmd.go]: With this check `operator-sdk new --vendor=true --dep-manager=dep` will result in the condition not being true and the error `--vendor cannot be used with --dep-manager=dep` will not be printed. 
Meaning `--vendor` can be used with `--dep-manager=dep`.
Unless I'm completely missing something?

Comment (estroz) [reply cmd/operator-sdk/new/cmd.go]: The error message should say `"--vendor cannot be set to false with --dep-manager=dep". On further consideration, we might want to skip checking `--vendor` entirely if `--dep-manager=dep` because dep requires `vendor`. WDYT?

Comment (joelanford) [reply cmd/operator-sdk/new/cmd.go]: > On further consideration, we might want to skip checking --vendor entirely if --dep-manager=dep

In that case `operator-sdk new --vendor=false --dep-manager=dep` would silently ignore `--vendor` and we'd create a vendor directory anyway. I think I prefer the message that says you can't disable vendor with dep.

Comment (LiliC) [reply cmd/operator-sdk/new/cmd.go]: I would not silently ignore it, but print out an error msg and still create the vendor, error msg something like:
` "dep requires vendor directory; ignoring vendor=false flag" `

File: cmd/operator-sdk/new/cmd.go [hunk 13]
```diff
@@ -60,6 +60,8 @@ generates a skeletal app-operator application in $GOPATH/src/github.com/example.
 	newCmd.Flags().StringVar(&depManager, "dep-manager", "modules", `Dependency manager the new project will use (choices: "dep", "modules")`)
 	newCmd.Flags().BoolVar(&skipGit, "skip-git-init", false, "Do not init the directory as a git repository")
 	newCmd.Flags().StringVar(&headerFile, "header-file", "", "Path to file containing headers for generated Go files. Copied to hack/boilerplate.go.txt")
+	newCmd.Flags().BoolVar(&makeVendor, "vendor", false, "Use a vendor directory for dependencies. This flag only applies when --dep-manager=modules (the default)")
+	newCmd.Flags().BoolVar(&skipValidation, "skip-validation", false, "Do not validate the resulting projects' structure and dependencies")
```
Comment (hasbro17) [inline cmd/operator-sdk/new/cmd.go]: Nit: There's only 1 project.
```suggestion
	newCmd.Flags().BoolVar(&skipValidation, "skip-validation", false, "Do not validate the resulting project's structure and dependencies")
```

File: doc/sdk-cli-reference.md [hunk 14]
```diff
@@ -263,6 +262,9 @@ Scaffolds a new operator project.
 * `--helm-chart-repo` string - Chart repository URL for the requested helm chart
 * `--helm-chart-version` string - Specific version of the helm chart (default is latest version)
 * `--dep-manager` string - Dependency manager the new project will use (choices: "dep", "modules") (default "modules")
+* `--skip-git-init` - Do not init the directory as a git repository
+* `--vendor` - Use a vendor directory for dependencies. This flag only applies when --dep-manager=modules (the default)
+* `--skip-validation` - Do not validate the resulting projects' structure and dependencies
```
Comment (hasbro17) [inline doc/sdk-cli-reference.md]: Nit: Singular
```suggestion
* `--skip-validation` - Do not validate the resulting project's structure and dependencies
```

File: doc/sdk-cli-reference.md [hunk 15]
```diff
@@ -263,6 +262,9 @@ Scaffolds a new operator project.
 * `--helm-chart-repo` string - Chart repository URL for the requested helm chart
 * `--helm-chart-version` string - Specific version of the helm chart (default is latest version)
 * `--dep-manager` string - Dependency manager the new project will use (choices: "dep", "modules") (default "modules")
+* `--skip-git-init` - Do not init the directory as a git repository
+* `--vendor` - Use a vendor directory for dependencies. This flag only applies when --dep-manager=modules (the default)
```
Comment (hasbro17) [inline doc/sdk-cli-reference.md]: Nit: Format like the other flag examples.
```suggestion
* `--vendor` - Use a vendor directory for dependencies. This flag only applies when `--dep-manager=modules` (the default)
```

File: doc/user-guide.md [hunk 16]
```diff
@@ -53,7 +53,9 @@ $ export GO111MODULE=on
 
 ##### Vendoring
 
-The Operator SDK uses [vendoring][go_vendoring] to supply dependencies to operator projects, regardless of the dependency manager. As with the above module mode constraint, we intend to allow use of dependencies [outside of `vendor`][module_vendoring] for projects in the near future.
+The Operator SDK uses [vendoring][go_vendoring] to supply dependencies to Go operator projects if the dependency manager is `modules` and the project was initialized with the `--vendor` flag set or if the dependency manager is `dep` (which always uses vendoring).
+
+By default, the operator's dependencies are managed with `modules` and `--vendor` is not set, so calls to `go {build,clean,get,install,list,run,test}` by `operator-sdk` subcommands will use an external modules directory. Execute `go help modules` for more information.
```
Comment (hasbro17) [inline doc/user-guide.md]: Can we move this before Line 56 so it's clear that the default behavior is `modules` with no vendoring?

And then we can state that vendoring can be used if XYZ.

Right now it's the caveats that are mentioned first.

File: internal/util/projutil/exec.go [hunk 17]
```diff
@@ -106,10 +119,19 @@ func getGeneralArgs(cmd string, opts GoCmdOptions) ([]string, error) {
 		if goModOn, err := GoModOn(); err != nil {
 			return nil, err
 		} else if goModOn {
-			bargs = append(bargs, "-mod=vendor")
+			// Does vendor exist?
+			info, err := os.Stat("vendor")
+			// Does the first "go" subcommand accept -mod=vendor?
+			_, ok := validVendorCmds[bargs[0]]
+			if err == nil && info.IsDir() && ok {
```
Comment (hasbro17) [inline internal/util/projutil/exec.go]: One last thing: `os.Stat("vendor")` could fail due to a permissions error, filesystem error, etc.
That means `vendor/` directory exists but with `err != nil` we silently treat it as if `vendor/` does not exist.

Instead we should explicitly return all other errors except `IsNotExist` error by first checking it with [`os.IsNotExist(err)`](https://golang.org/pkg/os/#IsNotExist):

```Go
info, err := os.Stat("vendor")
if err != nil && !os.IsNotExist(err) {
	return nil, err
}

_, ok := validVendorCmds[bargs[0]]
if err == nil && info.IsDir() && ok {
...
}
```

**Comment by estroz** [comment]
@hasbro17 we should test those cases, as well as `--dep-manager=dep`. I will write up a few PR's that add tests for builds.
